Sources/ApolloCodegenLib/Frontend/CompilationResult+SelectionSetMerging.swift

```diff
@@ -0,0 +1,97 @@
+import Foundation
+
+protocol SelectionMergable: ScopedSelectionSetHashable {
+  var _selectionSet:CompilationResult.SelectionSet? { get }
+  func merging(_: CompilationResult.SelectionSet) -> Self
+}
+
+extension CompilationResult.SelectionSet: SelectionMergable {
+
+  var _selectionSet: CompilationResult.SelectionSet? { self }
+
+  /// Returns a `Field` with the selections of the `newSelectionSet` merged in,
+  /// removing duplicates.
+  ///
+  /// - Note: If no changes were made the same `SelectionSet` is returned.
+  func merging(
+    _ newSelectionSet: CompilationResult.SelectionSet
+  ) -> Self {
+    let selectionsToMerge = newSelectionSet.selections.filter { !selections.contains($0) }
+
+    guard !selectionsToMerge.isEmpty else { return self }
+
+    let copy = self.copy()
+    copy.selections += selectionsToMerge
+    return copy
+  }
+
+  private func copy() -> Self {
+    return Self(parentType: self.parentType,
+                selections: self.selections)
+  }
+
+}
+
+extension CompilationResult.Field: SelectionMergable {
+
+  var _selectionSet: CompilationResult.SelectionSet? { self.selectionSet }
+
+  /// Returns a `Field` with the selections of the `newSelectionSet` merged in,
+  /// removing duplicates.
+  ///
+  /// - Note: If no changes were made the same `Field` is returned.
+  func merging(_ newSelectionSet: CompilationResult.SelectionSet) -> Self {
+    guard let existingSelectionSet = selectionSet else {
+      let copy = self.copy()
+      copy.selectionSet = newSelectionSet
+      return copy
+    }
+
+    let mergedSelectionSet = existingSelectionSet.merging(newSelectionSet)
+    guard mergedSelectionSet !== existingSelectionSet else { return self }
+
+    let copy = self.copy()
+    copy.selectionSet = mergedSelectionSet
+    return copy
+  }
+
+  private func copy() -> Self {
+    return Self(
+      name: self.name,
+      alias: self.alias,
+      arguments: self.arguments,
+      type: self.type,
+      selectionSet: self.selectionSet,
+      deprecationReason: self.deprecationReason,
+      description: self.description)
+  }
+
+}
+
+extension CompilationResult.Selection: SelectionMergable {
+
+  var _selectionSet: CompilationResult.SelectionSet? {
+    switch self {
+    case let .field(selection as SelectionMergable),
+      let .inlineFragment(selection as SelectionMergable):
+      return selection._selectionSet
+
+    case let .fragmentSpread(fragment):
+      return fragment.selectionSet
```

Comment:
why is this using the non-underscore `selectionSet` instead of the underscore one?

Reply:
`FragmentDefinition` doesn't conform to `SelectionMergable`, so it doesn't have the underscored version.

Theoretically, line 80 should currently never be called. Since we can't actually merge fragments. But if for some reason i the future I wanted to use this method to retrieve the selectionSet from a `SelectionMergable` entity in another context where it would be valid, then it will still jut work.

This is not the same as the fatalError in `merging(_ newSelectionSet)`. As doing that is not just reading a property, but is actually an operation that is non-sensical and shouldn't be allowed to occur on a fragment. If we call that on a fragment, we WANT to fail fast and loud so we can fix it.